File: Python/FirstBadVersion.py
# https://leetcode.com/problems/first-bad-version/


bad = 0
def isBadVersion(version: int) -> bool:
    if version >= bad:
        return True
    return False

def firstBadVersion(n: int) -> int:
    if n < 2:
        return 1
    
    left = 1
    right = n
    while (left <= right): 
        mid = (left + right) // 2
        if isBadVersion(mid) == True and isBadVersion(mid-1) == False:
            return mid 
        elif isBadVersion(mid-1) == True: 
            right = mid -1
        else: 
            left = mid + 1
    
bad = 4
print(firstBadVersion(5))

File: JavaScript/first-bad-version.js
var bad = 0

function isBadVersion(version){
    if(version >= bad){
        return true;
    } else {
        return false;
    }
}

function firstBadVersion(isBadVersion){
    return function(n){
        var start = 1;
        var end = n;
        var mid = parseInt(n/2);

        while(start<end){
            if(!isBadVersion(mid)){
                start = mid + 1;
            } else {
                end = mid
            }
            mid = parseInt((start + end) / 2);
        }
        return start;
    };
};


bad = 4 
console.log(firstBadVersion(5));